src/bsoncxx/types/value.cpp

```diff
@@ -40,15 +45,28 @@ value::value(const value& rhs) {
 }
 
 value& value::operator=(const value& rhs) {
-    _type = rhs._type;
+    if(_type && !rhs._type) {
```

Comment:
Also, I think we need a check for self-equality before doing anything here.  If `this == rhs` and we execute this code, we will get ourselves into a sticky situation by cleaning up our own union and then attempting to assign from it.